### PR TITLE
Porting NonlinearProgram to C++

### PR DESCRIPTION
**Pull request by crogers3**
This implementation of NonlinearProgram attempts to mirror the structure of `NonlinearProgram.m`. Currently, this implementation only supports snopt as the solver, and uses hardcoded resources.

### CODE REVIEW

**Comment by tkoolen** [comment]
It's awesome that you're doing this, but I have to say I really don't like the fact that the NonlinearProgram and Constraint classes depend on Snopt. These classes are going to be absolutely integral to Drake, and should in my opinion never depend on any particular solver, let alone an expensive commercial one.

**Comment by RussTedrake** [comment]
@tkoolen — don’t worry, i will resolve that.


On January 16, 2016 at 12:45:07 PM, Twan Koolen (notifications@github.com) wrote:

It's awesome that you're doing this, but I have to say I really don't like the fact that the NonlinearProgram and Constraint classes depend on Snopt. These classes are going to be absolutely integral to Drake, and should in my opinion never depend on any particular solver, let alone an expensive commercial one.

File: drake/solvers/CMakeLists.txt [hunk 1]
```diff
@@ -35,8 +35,14 @@ endif()
...
+  add_executable(TestNonlinearProgram NPCPP/testNonlinearProgram.cpp NPCPP/NonlinearProgram.cpp NPCPP/LinearConstraint.cpp NPCPP/QuadraticConstraint.cpp NPCPP/FunctionHandleConstraint NPCPP/BoundingBoxConstraint.cpp NPCPP/ConstantConstraint.cpp NPCPP/Constraint.cpp)
   pods_use_pkg_config_packages(NonlinearProgramSnoptmex snopt_c)
+  #pods_use_pkg_config_packages(TestSnoptC snopt_c)
+  pods_use_pkg_config_packages(TestNonlinearProgram snopt_c)
   set_target_properties(NonlinearProgramSnoptmex PROPERTIES COMPILE_FLAGS -fPIC)
+  #set_target_properties(TestSnoptC PROPERTIES COMPILE_FLAGS -fPIC)
+  set_target_properties(TestNonlinearProgram PROPERTIES COMPILE_FLAGS -fPIC)
 
   # link against matlab's copy of the gfortran library, if possible
   find_library(GFORTRAN_LIBRARY NAMES gfortran libgfortran
```
Comment (RussTedrake) [inline drake/solvers/CMakeLists.txt]: are we sure that we still need this?  snopt builds fine on systems that don't have fortran at all now (that is relatively recent), so I wouldn't think we'd actually need to link against it here..?  maybe i'm wrong.

File: drake/solvers/NPCPP/BoundingBoxConstraint.cpp [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#include "BoundingBoxConstraint.h"
```
Comment (RussTedrake) [inline drake/solvers/NPCPP/BoundingBoxConstraint.cpp]: I would tend to put many of the constraints into single files, Contraint.h and Constraint.cpp.  Splitting up files is good, but not when they're this simple??

File: drake/solvers/NPCPP/BoundingBoxConstraint.cpp [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+#include <utility>
+#include <vector>
+
+namespace drake {
+BoundingBoxConstraint::BoundingBoxConstraint(snopt::doublereal lb,
```
Comment (RussTedrake) [inline drake/solvers/NPCPP/BoundingBoxConstraint.cpp]: The constraint classes shouldn't know anything about snopt::doublereal.  They should just use doubles.  Only the solve method which is particular to snopt should use the snopt datatypes.  We will have other solvers without snopt that use these same constraint classes.

File: drake/solvers/NPCPP/NonlinearProgram.cpp [hunk 4]
```diff
@@ -0,0 +1,302 @@
+#include "NonlinearProgram.h"
```
Comment (RussTedrake) [inline drake/solvers/NPCPP/NonlinearProgram.cpp]: we're not going to have a nonlinear program class.  we want to build this into the optimizationproblem class as just a solve method.

**Comment by hongkai-dai** [comment]
@RussTedrake , do we still plan to implement `drakeFunction` first, and then just treat `constraint` as a container of `drakeFunction` and `bounds`? I feel that is a more cleaner implementation, as it allows nested functions.

**Comment by RussTedrake** [comment]
Yes, that’s the plan.

**Comment by hongkai-dai** [comment]
Then I think a starting point would be to define the `drakeFunction` class, and the hierarchy of this class first. The hierarchy would be a lot like that in `optimizationproblem` class, which supports `linearFunction`, `quadraticFunction`, `polynomialFunction`, etc. And it should support gradient computation, and probably sparsity pattern specification. I am not entirely sure how to handle integer and mixed integer constraint in `drakeFunction`. 

One thing I realized later that I should have supported in MATLAB `FunctionHandleConstraint` is that if a nonlinear function contains a linear part, such as `f(x,y) = g(x) + Ay`, then it is better to exploit this partial linear relationship, as it helps in the computation of Hessain. I hope we can do correct it this time in C++ implementation.

File: drake/solvers/CMakeLists.txt (hunk 1 above)
Comment (crogers3) [inline drake/solvers/CMakeLists.txt]: It looks like you are correct. I removed these links, and the code still built and ran properly. Good thing to know moving forward.

File: drake/solvers/NPCPP/BoundingBoxConstraint.cpp (hunk 2 above)
Comment (crogers3) [inline drake/solvers/NPCPP/BoundingBoxConstraint.cpp]: Done

File: drake/solvers/NPCPP/BoundingBoxConstraint.cpp (hunk 3 above)
Comment (crogers3) [inline drake/solvers/NPCPP/BoundingBoxConstraint.cpp]: Done

**Comment by crogers3** [comment]
Although the PR is closed, I've updated my branch with some of the comments from this thread, in case any of my code will carry over into the next iteration of this effort. I don't know if these commits will show in this closed PR though.

**Pull request closed without merging**